### PR TITLE
Support predefined jwt section as additional scope source (with additional cosmetics)

### PR DESCRIPTION
**Pull request by michaelklishin**
This is #40 by @papugamichal with various cosmetic changes. See #40 for background.

### CODE REVIEW

File: src/rabbit_auth_backend_oauth2.erl [hunk 1]
```diff
@@ -145,37 +150,75 @@ check_token(Token) ->
...
+            _ -> []
+            end,
+
+    case AdditionalScopes of
+        [] -> Payload;
+        _  ->
+            TokenScopes = maps:get(<<"scope">>, Payload),
+            case TokenScopes of
+                [] -> maps:put(<<"scope">>, AdditionalScopes, Payload);
+                _  -> maps:put(<<"scope">>, AdditionalScopes ++ TokenScopes, Payload)
```
Comment (acogoluegnes) [inline src/rabbit_auth_backend_oauth2.erl:205]: `scope` can still be [a binary or an array](#24) here, they're transformed in the last statement of `post_process_payload`. I'll add a test to see if this case is handled (complex claim is on and scope is a binary).

File: src/rabbit_auth_backend_oauth2.erl [hunk 2]
```diff
@@ -145,37 +150,75 @@ check_token(Token) ->
...
-    case Scopes of
-        [] ->
-            Payload;
+    case AdditionalScopes of
+        [] -> Payload;
         _  ->
-            maps:put(<<"scope">>, Scopes, Payload)
+            TokenScopes = maps:get(<<"scope">>, Payload),
+            case TokenScopes of
+                [] -> maps:put(<<"scope">>, AdditionalScopes, Payload);
```
Comment (acogoluegnes) [inline src/rabbit_auth_backend_oauth2.erl:220]: Same as above. Maybe the simplest would be to transform `scope` into an array at the beginning of `post_process_payload` and we should be fine afterwards.

Comment (acogoluegnes) [reply src/rabbit_auth_backend_oauth2.erl:220]: BTW, @papugamichal why do we change the Keycloak part? The initial request was `permissions` would replace `scope` alltogether and Keycloak did not complain about this implementation. What's the rationale here?

@michaelklishin WDYT?

**Review by acogoluegnes — COMMENTED** [review]
This looks OK overall, but needs a bit of rework. I'll polish the tests, add some, and check the system tests (not sure the environment changes are actually applied).

File: src/rabbit_auth_backend_oauth2.erl (hunk 2 above)
Comment (papugamichal) [reply src/rabbit_auth_backend_oauth2.erl:220]: @acogoluegnes previous implementation replaces scopes which came with JWT. Was it desirable behaviour?

File: src/rabbit_auth_backend_oauth2.erl (hunk 1 above)
Comment (papugamichal) [reply src/rabbit_auth_backend_oauth2.erl:205]: Fair point, i overlooked this case. Thank you!

File: src/rabbit_auth_backend_oauth2.erl (hunk 2 above)
Comment (papugamichal) [reply src/rabbit_auth_backend_oauth2.erl:220]: This is edge case - if token contains both ComplexClaims and Keycloack's scopes then post_process_payload_keycloak () will override existing ones (present in JWT) and those comming from post_process_payload_complex_claim(). IMO, this should be changed (see above implementation) or documentation should be updated to information about "side effects".

File: src/rabbit_auth_backend_oauth2.erl (hunk 1 above)
Comment (acogoluegnes) [reply src/rabbit_auth_backend_oauth2.erl:205]: I ended moving up the binary to array conversion, so complex scope and keycloak will always see an array downstream.

File: src/rabbit_auth_backend_oauth2.erl (hunk 2 above)
Comment (acogoluegnes) [reply src/rabbit_auth_backend_oauth2.erl:220]: Yes, the previous implementation would replace the scope alltogether. But I see your point: the complex scope processing could add some scopes that would then be erased by the keycloak processing. This should not happen the way you implemented it, this is defensive for an edge case, so we can keep it.

**Review by acogoluegnes — APPROVED** [review]

**Comment by michaelklishin** [comment]
Backported to `v3.8.x`.

**Comment by papugamichal** [comment]
@michaelklishin @acogoluegnes  thank you for review and support with handling this PR.

**Pull request merged**